adapters/Vungle/VungleAdapter/VungleRouter.m

```diff
@@ -65,33 +72,105 @@ - (void)initWithAppId:(NSString *)appId delegate:(id<VungleDelegate>)delegate {
+}
+
 - (void)addDelegate:(id<VungleDelegate>)delegate {
-  @synchronized(self.delegates) {
-    [_delegates setObject:delegate forKey:delegate.desiredPlacement];
+  if (delegate.adapterAdType == Interstitial || delegate.adapterAdType == Rewarded) {
+    @synchronized(self.delegates) {
+      if (delegate && ![self.delegates objectForKey:delegate.desiredPlacement]) {
+        [self.delegates setObject:delegate forKey:delegate.desiredPlacement];
+      }
+    }
+  } else if (delegate.adapterAdType == MREC) {
+    @synchronized(self.bannerDelegates) {
+      self.mrecPlacementID = [delegate.desiredPlacement copy];
+      if (delegate && ![self.bannerDelegates objectForKey:delegate.desiredPlacement]) {
+        [self.bannerDelegates setObject:delegate forKey:delegate.desiredPlacement];
+      }
+    }
   }
 }
 
 - (id<VungleDelegate>)getDelegateForPlacement:(NSString *)placement {
-  @synchronized(self.delegates) {
-    return [_delegates objectForKey:placement];
+  id<VungleDelegate> delegate;
+  if ([placement isEqualToString:self.mrecPlacementID]) {
+    @synchronized(self.bannerDelegates) {
+      if ([self.bannerDelegates objectForKey:placement]) {
+        delegate = [self.bannerDelegates objectForKey:placement];
+        if (delegate.bannerState != BannerRouterDelegateStateRequesting) {
+          delegate = nil;
+        }
+      }
+    }
+  } else {
+    @synchronized(self.delegates) {
+      if ([self.delegates objectForKey:placement]) {
+        delegate = [self.delegates objectForKey:placement];
+      }
+    }
   }
+
+  return delegate;
 }
 
-- (void)removeDelegateForPlacementID:(NSString *)placementID {
-  @synchronized(self.delegates) {
-    [_delegates removeObjectForKey:placementID];
+- (void)removeDelegate:(id<VungleDelegate>)delegate {
+  if (delegate.adapterAdType == Interstitial || delegate.adapterAdType == Rewarded) {
+    @synchronized(self.delegates) {
+      if (delegate && [self.delegates objectForKey:delegate.desiredPlacement]) {
+        [self.delegates removeObjectForKey:delegate.desiredPlacement];
+      }
+    }
+  } else if (delegate.adapterAdType == MREC) {
+    @synchronized(self.bannerDelegates) {
+      if (delegate && [self.bannerDelegates objectForKey:delegate.desiredPlacement]) {
+        [self.bannerDelegates removeObjectForKey:delegate.desiredPlacement];
+      }
+    }
   }
 }
 
+- (BOOL)hasDelegateForPlacementID:(NSString *)placementID
+                      adapterType:(VungleNetworkAdapterAdType)adapterType {
+  BOOL result = NO;
+  NSMapTable<NSString *, id<VungleDelegate>> *delegates;
+  if ([self isSDKInitialized]) {
+    delegates = [self.delegates copy];
+  } else {
+    delegates = [self.initializingDelegates copy];
+  }
+
+  for (NSString *key in delegates) {
+    id<VungleDelegate> delegate = [delegates objectForKey:key];
+    if (delegate.adapterAdType == adapterType &&
+        [delegate.desiredPlacement isEqualToString:placementID]) {
+      result = YES;
```

Comment:
Can we `return YES` from here and `NO` at the end so there is no need to use a `result` boolean?

Reply:
@joshuajplagonera Great suggestion, thanks. :)

Reply:
@joshuajplagonera Thanks, done. 😊